### PR TITLE
LE Audio: CSIS Ordered Access procedure 

### LINKED ISSUE #42341: LE Audio: CSIS Ordered Access procedure use rank
**Is your enhancement proposal related to a problem? Please describe.**
The ordered access procedure has mostly been implemented by `bt_csis_client_get_lock_state`, but that only reads the lock states of the members by rank, but does not execute any procedure. The point of the procedure is that we perform a specific procedure on each device in ranked order. 

**Describe the solution you'd like**
After reading the lock states, the procedure shall be executed on each device in ranked order. The only way to provide a procedure is by callback, so a callback should be defined and used in a new function that internally calls `bt_csis_client_get_lock_state`, that is called for each set member. 

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by Thalley**
Updates the API and implementation to implement the CSIP ordered access procedure. 

fixes https://github.com/zephyrproject-rtos/zephyr/issues/42341

### CODE REVIEW

File: include/bluetooth/audio/csis.h [hunk 1]
```diff
@@ -282,20 +282,22 @@ typedef void (*bt_csis_client_lock_changed_cb)(struct bt_csis_client_csis_inst *
 					       bool locked);
 
 /**
- * @typedef bt_csis_client_lock_state_read_cb
- * @brief Callback for bt_csis_client_get_lock_state()
+ * @typedef bt_csis_client_ordered_access_cb_t
+ * @brief Callback for bt_csis_client_ordered_access()
```
Comment (MariuszSkamra) [inline include/bluetooth/audio/csis.h:286]: If the callback is used from `bt_csis_client_ordered_access` context, why it's lifetime is different? I.e. why it's `bt_csis_client_cb` member?

File: subsys/bluetooth/audio/csis_client.c [hunk 2]
```diff
@@ -1412,26 +1559,31 @@ int bt_csis_client_lock(const struct bt_csis_client_set_member **members,
...
 	}
 
-	cur_inst = NULL;
-	err = verify_members_and_get_inst(members, count, set_info, true, &cur_inst);
+	err = verify_members(members, count, set_info);
 	if (err != 0) {
 		BT_DBG("Could not verify members: %d", err);
 		return err;
 	}
+	active_members_store_ordered(members, count, set_info, true);
```
Comment (MariuszSkamra) [inline subsys/bluetooth/audio/csis_client.c:1567]:
```suggestion

	active_members_store_ordered(members, count, set_info, true);
```

new empty line

File: subsys/bluetooth/audio/csis_client.c [hunk 3]
```diff
@@ -1381,27 +1505,50 @@ int bt_csis_client_get_lock_state(const struct bt_csis_client_set_member **membe
...
```
Comment (MariuszSkamra) [inline subsys/bluetooth/audio/csis_client.c:1513]: ditto

File: subsys/bluetooth/audio/csis_client.c [hunk 4]
```diff
@@ -140,6 +143,110 @@ static struct bt_csis *lookup_instance_by_set_info(const struct bt_csis_client_s
...
+	}
+
+	if (inst_1->cli.rank < inst_2->cli.rank) {
+		return 1;
+	} else if (inst_1->cli.rank > inst_2->cli.rank) {
+		return -1;
+	}
+
+	return 0;
+}
```
Comment (MariuszSkamra) [inline subsys/bluetooth/audio/csis_client.c:212]: Are those separate functions needed just to return -1 vs 1? Please extract common part

File: subsys/bluetooth/audio/csis_client.c [hunk 5]
```diff
@@ -140,6 +143,110 @@ static struct bt_csis *lookup_instance_by_set_info(const struct bt_csis_client_s
...
+	}
+
+	if (inst_2 == NULL) {
+		BT_ERR("inst_2 was NULL for member %p", member_2);
+		return 0;
+	}
+
+	if (inst_1->cli.rank < inst_2->cli.rank) {
+		return 1;
+	} else if (inst_1->cli.rank > inst_2->cli.rank) {
```
Comment (MariuszSkamra) [inline subsys/bluetooth/audio/csis_client.c:207]: else not needed as you return from if

File: subsys/bluetooth/audio/csis_client.c [hunk 6]
```diff
@@ -140,6 +143,110 @@ static struct bt_csis *lookup_instance_by_set_info(const struct bt_csis_client_s
...
+			const uint8_t rank_2 = inst_2->cli.rank;
+
+			if (ascending) {
+				__ASSERT(rank_1 <= rank_2,
+					 "Members not sorted by ascending rank %u - %u",
+					 rank_1, rank_2);
+			} else {
+				__ASSERT(rank_1 >= rank_2,
+					"Members not sorted by descending rank %u - %u",
+					rank_1, rank_2);
```
Comment (MariuszSkamra) [inline subsys/bluetooth/audio/csis_client.c:244]: Is this assert needed? Couldn't be verified by a bsim_test or unit test?

**Review by MariuszSkamra — CHANGES_REQUESTED** [review]
Some comments. Looking at the code I see that there is a static structure `active_members` that holds context. I guess the user might get into trouble when tries to use this API with multithreads and forbids concurrent operations on 2 sets.

File: tests/bluetooth/bsim_bt/bsim_test_audio/src/csis_client_test.c [hunk 7]
```diff
@@ -70,21 +71,18 @@ static void csis_lock_changed_cb(struct bt_csis_client_csis_inst *inst,
...
-	if (locked) {
-		set_read_locked = true;
+	if (err) {
+		FAIL("Ordered access failed with err %d\n", err);
+	} else if (locked) {
+		printk("Ordered access procedure locked member %p\n", member);
+		ordered_access_locked = true;
 	} else {
-		set_read_unlocked = true;
+		printk("Ordered access procedure finished\n");
```
[the remark this answers is not in the document]
Comment (Thalley) [reply tests/bluetooth/bsim_bt/bsim_test_audio/src/csis_client_test.c:84]: `ordered_access_finished` was actually unused, so removed it.

File: tests/bluetooth/bsim_bt/bsim_test_audio/src/csis_client_test.c [hunk 8]
```diff
@@ -185,38 +196,40 @@ static void discover_members_timer_handler(struct k_work *work)
...
 		return;
 	}
 
 	if (expect_locked) {
-		WAIT_FOR(set_read_locked);
+		WAIT_FOR(ordered_access_locked);
 	} else {
-		WAIT_FOR(set_read_unlocked);
+		WAIT_FOR(ordered_access_unlocked);
 	}
```
[the remark this answers is not in the document]
Comment (Thalley) [reply tests/bluetooth/bsim_bt/bsim_test_audio/src/csis_client_test.c:225]: `ordered_access_finished` was actually unused, so removed it.

File: subsys/bluetooth/audio/csis_client.c [hunk 9]
```diff
@@ -1025,11 +1109,22 @@ static void csis_client_write_lock_cb(struct bt_conn *conn, uint8_t err,
...
+
+			active_members_reset();
+			if (csis_client_cbs != NULL &&
+			csis_client_cbs->lock_set != NULL) {
+				csis_client_cbs->lock_set(-ENOENT);
+			}
+		}
 
 		csis_client_err = csis_client_write_set_lock(cur_inst, true,
 							     csis_client_write_lock_cb);
```
[the remark this answers is not in the document]
Comment (Thalley) [reply subsys/bluetooth/audio/csis_client.c:1130]: I removed a bit of the duplicated code, but some still remain as that is not easily moved to a different function IMO.

File: include/bluetooth/audio/csis.h [hunk 10]
```diff
@@ -327,21 +329,40 @@ bool bt_csis_client_is_set_member(uint8_t set_sirk[BT_CSIS_SET_SIRK_SIZE],
...
+ *                   done on the members in ascending order.
+ * @param count      Number of members in @p members.
  *
- * @param members   Array of set members to check lock state for.
+ * @return true if the procedures can be successfully done, or false to stop the
+ *         procedure.
+ */
+typedef bool (*bt_csis_client_ordered_access_t)(const struct bt_csis_client_set_info *set_info,
+						struct bt_csis_client_set_member *members[],
+						size_t count);
```
[the remark this answers is not in the document]
Comment (Thalley) [reply include/bluetooth/audio/csis.h:344]: I agree with all your points. 

> First of all, the bt_csis_client_ordered_access says that cb is called in the order of the member's ranks (if present) but the the callback does not have a member parameter, but the members (ordered by rank I suppose).
Thus, shouldn't this callback look like this ?

After a discussion with Asbjørn I changed it from a foreach to a single callback with all members. That seemed easier to use for all parties. 

> Secondly, you add another callback function named bt_csis_client_ordered_access_cb_t. That is confusing having those two with so similar names.

> Thirdly, Is it really needed to have 2 callbacks for this procedure? I have a feeling that the one given to bt_csis_client_ordered_access with additional err parameter (to indicate that procedure cannot be performed because one of the peer's is locked) should be sufficient.

One callback is called when the procedure is ready to be done on each member, and the other is when the ordered access procedure is completed. 

Ideally I'd like to reduce this to a single callback, but the parameters in each callback are sufficiently different that it doesn't really make sense IMO. 

I do, however, want to find a better name for these, as `_cb_` is the only difference between them... 

Basically we have
```c
bt_csis_client_ordered_access()
bt_csis_client_ordered_access_t() // callback
bt_csis_client_ordered_access_cb_t() // callback for bt_csis_client_ordered_access
```

Personally I think a new name for `bt_csis_client_ordered_access_t` is the best solution here. 

How do you feel about `bt_csis_client_ordered_access_ready_t` or `bt_csis_client_ordered_access_execute_t`? 
Do you have any other ideas, or perhaps an idea to reduce the callbacks to a single callback?

File: include/bluetooth/audio/csis.h (hunk 1 above)
Comment (Thalley) [reply include/bluetooth/audio/csis.h:286]: Are you asking why this callback is not provided as an argument to `bt_csis_client_ordered_access` instead? 

That is certainly a possibility, but the `bt_csis_client_ordered_access_t` is more likely to change between each call to `bt_csis_client_ordered_access`, whereas `bt_csis_client_ordered_access_cb_t` is more likely the remain the same function. 

e.g. if you want to execute two procedures on an array of set members, you would call `bt_csis_client_ordered_access` with two different callbacks.

File: subsys/bluetooth/audio/csis_client.c (hunk 4 above)
Comment (Thalley) [reply subsys/bluetooth/audio/csis_client.c:212]: The only thing I can see that I can make common is moving the casting of `const void *` to `struct bt_csis_client_set_member *`, which would require me to create a function that takes a `const void *` that then casts this to a `struct bt_csis_client_set_member *` and then does the call to `lookup_instance_by_set_info`. 

But that extra function would just remove two lines 
```
	struct bt_csis_client_set_member *member_1 = *(struct bt_csis_client_set_member **)m1;
	struct bt_csis_client_set_member *member_2 = *(struct bt_csis_client_set_member **)m2;
```
and replace 
```c
	inst_1 = lookup_instance_by_set_info(member_1, active.info);
```
with something like
```c
	inst_1 = member_rank_compare_lookup_instance_by_set_info(m1, active.info);
```

I wanted to use `qsort_r` which allows for an additional argument, which could also make this a single function instead, but that is not POSIX compliant, and won't work with e.g. babblesim nor native_posix. 
The alternative is to store the comparison direction (ascending/descending) in the `active` struct. 

The last alternative is to create a new struct that holds a tuple of `struct bt_csis *` and return that. That way it could made common. 

If you want, I can create the tuple struct return type, and extract common code, or I can add a sorting direction variable in `active`, which would reduce this to a single function. Which do you prefer?

File: subsys/bluetooth/audio/csis_client.c (hunk 5 above)
Comment (Thalley) [reply subsys/bluetooth/audio/csis_client.c:207]: The else covers the case where `inst_1->cli.rank == inst_2->cli.rank`

File: subsys/bluetooth/audio/csis_client.c (hunk 6 above)
Comment (Thalley) [reply subsys/bluetooth/audio/csis_client.c:244]: The application does have access to the rank value at this point, so it won't be able to do such as verification (actually tried to implement this in the babblesim tests first time around).

File: subsys/bluetooth/audio/csis_client.c (hunk 5 above)
Comment (MariuszSkamra) [reply subsys/bluetooth/audio/csis_client.c:222]: I know, I was thinking rather of:
``
if (a > b) {
  return 1;
}

if (a < b) {
  return -1;
}

return 0;
```

Comment (Thalley) [reply subsys/bluetooth/audio/csis_client.c:222]: Right. 

Are you aware of any branching optimizations or downsides of doing
```c
if (a > b} {
  return 1;
}

if (b < a) {
  return -1;
}

return 0;
```

compared to 
```c
if (a > b} {
  return 1;
} else if (b < a) {
  return -1;
}

return 0;
```

I can change it no problem, even if it does not really change anything. 


With regards to doing `return inst_2->cli.rank - inst_1->cli.rank`, the `qsort` implementation does to provide enough information about whether the value can be less than -1 or greater than 1, but from the Linux `man`page:

> The comparison function must return an integer less than, equal to, or greater  than  zero if  the  first  argument  is considered to be respectively less than, equal to, or greater than the second.  If two members compare as equal, their order in the sorted array is  undefined.

So I guess I can just do the simple subtraction - Will update.

**Comment by Thalley** [comment]
Rebased to solve merge conflicts

File: include/bluetooth/audio/csis.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (MariuszSkamra) [reply include/bluetooth/audio/csis.h]: > After a discussion with Asbjørn I changed it from a foreach to a single callback with all members. That seemed easier to use for all parties.

OK, but the documentation was out of date I guess. Thanks for update ^^ Maybe it would be nice to have a info that `@ref bt_csis_client_ordered_access_cb_t` will be called when procedure is completed.

File: include/bluetooth/audio/csis.h (hunk 1 above)
Comment (MariuszSkamra) [reply include/bluetooth/audio/csis.h]: And then how do you know which procedure is complete?

File: subsys/bluetooth/audio/csis_client.c (hunk 4 above)
Comment (MariuszSkamra) [reply subsys/bluetooth/audio/csis_client.c]: Try this one:
```
static int member_rank_compare_desc(const void *m1, const void *m2)
{
    return member_rank_compare_asc(m2, m1);
}
```

File: subsys/bluetooth/audio/csis_client.c (hunk 6 above)
Comment (MariuszSkamra) [reply subsys/bluetooth/audio/csis_client.c:241]: My point is that you sort the members, and then try to assert if not sorted. I'm OK with asserting things that one is not 100% certain, but here...

File: include/bluetooth/audio/csis.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (Thalley) [reply include/bluetooth/audio/csis.h]: Added that in the documentation now as well :)

File: include/bluetooth/audio/csis.h (hunk 1 above)
Comment (Thalley) [reply include/bluetooth/audio/csis.h]: Hmm, `bt_csis_client_ordered_access` could in theory be called with the same array of members and `set_info` at the same time (currently that would fail due to the `busy` flag), and then the callback would be called with the same parameters, yes. 

Currently (due to the `busy` flag) it is not a problem, but I see what you mean. 

Are you suggesting that we switch to use parameter-based callback for all client operations callbacks (like the GATT API)?

File: subsys/bluetooth/audio/csis_client.c (hunk 4 above)
Comment (Thalley) [reply subsys/bluetooth/audio/csis_client.c]: That's a great suggestion - Did not think of that myself :D

File: subsys/bluetooth/audio/csis_client.c (hunk 6 above)
Comment (Thalley) [reply subsys/bluetooth/audio/csis_client.c:241]: The solutions I see are
1) (current) Do an assert to verify that the sorting functionality works
2) Have a babblesim test use the internal header file and access the rank
3) Expose the rank to the application
4) Replace the asserts with a `BT_ERR`

File: include/bluetooth/audio/csis.h (hunk 1 above)
Comment (MariuszSkamra) [reply include/bluetooth/audio/csis.h]: I guess that way the context would not be lost. From the API user perspective this would be used like this pseudo code:
```
static screen_indicator volume_indicator;

void group_volume_up_complete(group, err)
{
  screen_indicate(screen, screen_indicator, indicate_vol_up);
}

int group_volume_up_func(group/group_member)
{
   return volume_up(group_member);
}

int group_volume_up(group)
{
    return bt_csis_client_ordered_access(group, group_volume_up_func, group_volume_up_complete);
}
```

Comment (Thalley) [reply include/bluetooth/audio/csis.h]: Right. 

I see what you mean, and I see the value of it, especially if we want to enable parallel operations (I don't recall the CSIP specifying that you can't do multiple ordered access procedures in parallel (of course, the "correct" way of doing this would be do do multiple things in the first callback, instead of doing the Ordered Access Procedure multiple times in parallel). 

However, the points you've made here is much more general than just this procedure, and generally applies to all client procedures.

Comment (MariuszSkamra) [reply include/bluetooth/audio/csis.h]: I've created an issue here: https://github.com/zephyrproject-rtos/zephyr/issues/44867

File: subsys/bluetooth/audio/csis_client.c [hunk 11]
```diff
@@ -140,6 +143,90 @@ static struct bt_csis *lookup_instance_by_set_info(const struct bt_csis_client_s
...
+{
+	(void)memcpy(active.members, members, count * sizeof(members[0]));
+	active.members_count = count;
+	active.info = info;
+
+	qsort(active.members, count, sizeof(members[0]),
+	      ascending ? member_rank_compare_asc : member_rank_compare_desc);
+
+	if (IS_ENABLED(CONFIG_ASSERT) &&
+	    IS_ENABLED(CONFIG_BT_DEBUG_CSIS_CLIENT)) {
```
Comment (MariuszSkamra) [inline subsys/bluetooth/audio/csis_client.c:207]: I guess `CONFIG_BT_DEBUG_CSIS_CLIENT` is undesirable here, as when `CONFIG_ASSERT` is enabled you might want to assert, no matter if CSIS debug is enabled.

File: subsys/bluetooth/audio/csis_client.c (hunk 6 above)
Comment (MariuszSkamra) [reply subsys/bluetooth/audio/csis_client.c:224]: If the check is necessary, I guess the assert is the best option from those you listed

File: subsys/bluetooth/audio/csis_client.c (hunk 11 above)
Comment (Thalley) [reply subsys/bluetooth/audio/csis_client.c]: Good point. This was a leftover from when I just did a `BT_ERR` instead.

**Comment by Thalley** [comment]
Fixes CI issue with array out of bounds (fixed by checking if `CONFIG_BT_MAX_CONN > 1`

**Review by MariuszSkamra — DISMISSED** [review]

File: subsys/bluetooth/audio/csis_client.c [hunk 12]
```diff
@@ -140,6 +143,90 @@ static struct bt_csis *lookup_instance_by_set_info(const struct bt_csis_client_s
...
+
+static int member_rank_compare_desc(const void *m1, const void *m2)
+{
+	/* If we call the "compare ascinding" function with the members
+	 * reversed, it will work as a descending comparison
+	 */
+	return member_rank_compare_asc(m2, m1);
+}
+
+static void active_members_store_ordered(struct bt_csis_client_set_member *members[],
```
Comment (larsgk) [inline subsys/bluetooth/audio/csis_client.c:194]: Looking through the PR, it doesn't seem like the original array of pointers is modified.  Is there a reason for this not being const?

Comment (larsgk) [inline subsys/bluetooth/audio/csis_client.c:194]: Looking through the PR, it doesn't seem like the original array of pointers is modified.  Is there a reason for this not being const?

Comment (Thalley) [reply subsys/bluetooth/audio/csis_client.c:194]: Essentially it is because `bt_csis_client_set_member` contains the `bt_conn`, and all the lower layer APIs (including GATT) does not have the `bt_conn` parameter as const. 

Besides that, we also have that `bt_csis_client_set_member.insts[x].csis` that also need to be non-const as that will also change during the procedures. 

Some of this API is changed again in https://github.com/zephyrproject-rtos/zephyr/pull/45164 to accommodate the CAP initiator, so some of this will unfortunately be changed again.

Comment (larsgk) [reply subsys/bluetooth/audio/csis_client.c:194]: ok, thanks - LGTM then

**Review by larsgk — APPROVED** [review]
LGTM

**Review by larsgk — DISMISSED** [review]

**Comment by Thalley** [comment]
Rebased to solve merge conflicts

**Review by MariuszSkamra — DISMISSED** [review]

File: subsys/bluetooth/audio/csis_client.c [hunk 13]
```diff
@@ -137,6 +140,90 @@ static struct bt_csis *lookup_instance_by_set_info(const struct bt_csis_client_s
...
+{
+	(void)memcpy(active.members, members, count * sizeof(members[0U]));
+	active.members_count = count;
+	active.info = info;
+
+	if (count > 1U && CONFIG_BT_MAX_CONN > 1) {
+		qsort(active.members, count, sizeof(members[0U]),
+		ascending ? member_rank_compare_asc : member_rank_compare_desc);
+
+		if (IS_ENABLED(CONFIG_ASSERT) && CONFIG_BT_MAX_CONN > 1) {
```
Comment (sjanc) [inline subsys/bluetooth/audio/csis_client.c:204]: this is already underl BT_MAX_CON>1

**Review by sjanc — DISMISSED** [review]

**Review by sjanc — DISMISSED** [review]

**Review by fredrikdanebjer — DISMISSED** [review]

File: subsys/bluetooth/audio/csis_client.c [hunk 14]
```diff
@@ -183,43 +185,46 @@ static int member_rank_compare_asc(const void *m1, const void *m2)
...
 {
-	struct bt_csis_client_set_member *member_1 = *(struct bt_csis_client_set_member **)m1;
-	struct bt_csis_client_set_member *member_2 = *(struct bt_csis_client_set_member **)m2;
-	struct bt_csis *inst_1;
-	struct bt_csis *inst_2;
-
-	/* TODO: lookup_instance_by_set_info should be able to work on a const
-	 * member, but it relies on `bt_conn_index` which doesn't take a const
-	 * bt_conn.
+	/* If we call the "compare ascinding" function with the members
```
Comment (asbjornsabo) [inline subsys/bluetooth/audio/csis_client.c:185]: ascinding -> ascending

Neat simplification, though.  Could this rather be done in the previous commit, where the functions are introduced?

**Review by asbjornsabo — CHANGES_REQUESTED** [review]
Some minor things.
Also: There seems to be some overlap between the second and third commit - with some code introduced in the second commit being changed again in the third.  I think it would simplify the commits if the "net result" was introduced in the second commit.

File: include/zephyr/bluetooth/audio/csis.h [hunk 15]
```diff
@@ -282,20 +282,22 @@ typedef void (*bt_csis_client_lock_changed_cb)(struct bt_csis_client_csis_inst *
...
- * @brief Callback for bt_csis_client_get_lock_state()
+ * @typedef bt_csis_client_ordered_access_cb_t
+ * @brief Callback for bt_csis_client_ordered_access()
  *
- * If any of the set members supplied to bt_csis_client_get_lock_state() is
- * in the locked state, this will be called with @p locked true. If any
- * set member is in the locked state, the remaining (if any) won't be read.
+ * If any of the set members supplied to bt_csis_client_ordered_access() is
+ * in the locked state, this will be called with @p locked true and @p member
+ * will the locked member, and the ordered access procedure is cancelled.
```
[the remark this answers is not in the document]
Comment (Thalley) [reply include/zephyr/bluetooth/audio/csis.h:290]: Added a `be`

File: subsys/bluetooth/audio/csis_client.c [hunk 16]
```diff
@@ -53,6 +53,8 @@ static struct active_members {
 	uint8_t members_count;
 	uint8_t members_handled;
 	uint8_t members_restored;
+
```
[the remark this answers is not in the document]
Comment (Thalley) [reply subsys/bluetooth/audio/csis_client.c:56]: To separate the fields; `oap_cb` is not a field that represent the members or any of the counters, but rather a single value for the procedure itself.

**Comment by Thalley** [comment]
@asbjornsabo w.r.t doing things in earlier commits: Yes, it is always possible to do the change in the previous commit, but it kind of goes against does a single thing in a commit (to keep commits as simple as possible). I'd rather keep it as is.

**Comment by asbjornsabo** [comment]
> @asbjornsabo w.r.t doing things in earlier commits: Yes, it is always possible to do the change in the previous commit, but it kind of goes against does a single thing in a commit (to keep commits as simple as possible). I'd rather keep it as is.

I was thinking of the member_rank_compare_asc/desc - they were introduced in one commit, and then the "desc" was changed in the next one to just call the "asc". I think it would have been simpler to just introduce them in the final form in the first commit instead of spreading it over two.  IMO, that would be a single thing.
But I am OK to keep as is now.

**Review by asbjornsabo — APPROVED** [review]
My comments have been addressed.

**Review by sjanc — APPROVED** [review]
